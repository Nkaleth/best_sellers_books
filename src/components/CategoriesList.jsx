import { useSelector, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import { NavLink } from 'react-router-dom';
import Categorie from './Categorie';
import { getCategoriesList, filterBySearch } from '../redux/categories/categoriesSlice';
import Navbar from './Navbar';
import SearchBox from './SearchBox';
import '../styles/categoriesList.css';
import search from '../icons/search-svgrepo-com.svg';

const CategoriesList = () => {
  const {
    categoriesList,
    isLoading,
    categoriesFiltered,
  } = useSelector((store) => store.categories);
  const dispatch = useDispatch();
  useEffect(() => {
    if (categoriesList.length === 0) { dispatch(getCategoriesList()); }
  });

  useEffect(() => {
    dispatch(filterBySearch(''));
  }, [dispatch]);

  if (isLoading) {
    return (
      <div className="loading">
        <h1>Loading...</h1>
      </div>
    );
  }
  if (categoriesFiltered.length > 0) {
    return (
      <>
        <Navbar categorie="Best Sellers" numCateg={categoriesList.length} />
        <section className="iconBox">
          <SearchBox />
          <img src={search} alt="findIcon" />
        </section>
        <p className="titleBoxCategories">BOOKS BY CATEGORY</p>
        <section className="CategoriesBox">
          { categoriesFiltered.map((categorie) => (
            <NavLink className="categorieLink" key={categorie.list_name_encoded} to={`/details/${categorie.list_name_encoded}`}>
              <Categorie
                id={categorie.list_name_encoded}
                categorieName={categorie.display_name}
                date={categorie.newest_published_date}
              />
            </NavLink>
          ))}
        </section>
      </>
    );
  }
  return (
    <>
      <Navbar categorie="BEST SELLERS" numCateg={categoriesList.length} />
      <section className="iconBox">
        <SearchBox />
        <img src={search} alt="findIcon" />
      </section>
      <p>BOOKS BY CATEGORY</p>
      <section className="CategoriesBox">
        { categoriesList.map((categorie) => (
          <NavLink className="categorieLink" key={categorie.list_name_encoded} to={`/details/${categorie.list_name_encoded}`}>
            <Categorie
              id={categorie.list_name_encoded}
              categorieName={categorie.display_name}
              date={categorie.newest_published_date}
              index={categorie.newindex}
            />
          </NavLink>
        ))}
      </section>
    </>
  );
};
export default CategoriesList;
