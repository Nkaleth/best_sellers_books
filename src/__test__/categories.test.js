import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import CategoriesList from '../components/CategoriesList';

describe('Testing Home Page: <CategoriesList /> component', () => {
  const initialState = {
    categoriesList: [{
      list_name: 'Combined Print and E-Book Fiction',
      display_name: 'Combined Print & E-Book Fiction',
      list_name_encoded: 'combined-print-and-e-book-fiction',
      oldest_published_date: '2011-02-13',
      newest_published_date: '2023-03-19',
      updated: 'WEEKLY',
    },
    {
      list_name: 'Combined Print and E-Book Nonfiction',
      display_name: 'Combined Print & E-Book Nonfiction',
      list_name_encoded: 'combined-print-and-e-book-nonfiction',
      oldest_published_date: '2011-02-13',
      newest_published_date: '2023-03-19',
      updated: 'WEEKLY',
    },
    {
      list_name: 'Hardcover Fiction',
      display_name: 'Hardcover Fiction',
      list_name_encoded: 'hardcover-fiction',
      oldest_published_date: '2008-06-08',
      newest_published_date: '2023-03-19',
      updated: 'WEEKLY',
    }],
    categoriesFiltered: [],
    isLoading: false,
  };

  const categoriesSlice = createSlice({
    name: 'categories',
    initialState,
  });

  const store = configureStore({
    reducer: {
      categories: categoriesSlice.reducer,
    },
  });

  const wrapper = (component) => (
    <Provider store={store}>
      <BrowserRouter>
        {component}
      </BrowserRouter>
    </Provider>
  );

  it("Should render 'books by category' text", () => {
    render(wrapper(<CategoriesList />));
    expect(screen.getByText(/books by category/i)).toBeInTheDocument();
  });

  it('Should render 3 categories', () => {
    render(wrapper(<CategoriesList />));
    expect(screen.getAllByTestId('categorieBox')).toHaveLength(3);
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(<CategoriesList />)).toJSON()).toMatchSnapshot();
  });
});
