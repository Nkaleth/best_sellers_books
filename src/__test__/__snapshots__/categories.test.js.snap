// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home Page: <CategoriesList /> component Should match the snapshot 1`] = `
Array [
  <nav
    className="navbarTop"
  >
    <a
      aria-current="page"
      className="active"
      href="/"
      onClick={[Function]}
    >
      &lt;
    </a>
    <p>
      2023
       
    </p>
    <p>
      best sellers books
    </p>
    <img
      alt="mic"
      src=""
    />
    <img
      alt="config"
      src=""
    />
  </nav>,
  <section
    className="navbarBottom"
  >
    <div
      className="catImgContainer"
    >
      <img
        alt="catImg"
        src=""
      />
    </div>
    <div>
      <h1
        className="titleNavbar"
      >
        Best Sellers
      </h1>
      <p
        className="counterNavbar"
      >
        3
         
        Categories
      </p>
    </div>
  </section>,
  <input
    onChange={[Function]}
    placeholder="Search by Category"
    type="text"
  />,
  <p>
    BOOKS BY CATEGORY
  </p>,
  <section
    className="CategoriesBox"
  >
    <a
      className=""
      href="/details/combined-print-and-e-book-fiction"
      onClick={[Function]}
    >
      <section
        className="categorieBox"
        data-testid="categorieBox"
        id="combined-print-and-e-book-fiction"
      >
        <div
          className="categorieImageBox"
        >
          <img
            alt="Categorie Icon"
            src=""
          />
        </div>
        <div>
          <h2>
            Combined Print & E-Book Fiction
          </h2>
          <p>
            Published date:
            2023-03-19
          </p>
        </div>
      </section>
    </a>
    <a
      className=""
      href="/details/combined-print-and-e-book-nonfiction"
      onClick={[Function]}
    >
      <section
        className="categorieBox"
        data-testid="categorieBox"
        id="combined-print-and-e-book-nonfiction"
      >
        <div
          className="categorieImageBox"
        >
          <img
            alt="Categorie Icon"
            src=""
          />
        </div>
        <div>
          <h2>
            Combined Print & E-Book Nonfiction
          </h2>
          <p>
            Published date:
            2023-03-19
          </p>
        </div>
      </section>
    </a>
    <a
      className=""
      href="/details/hardcover-fiction"
      onClick={[Function]}
    >
      <section
        className="categorieBox"
        data-testid="categorieBox"
        id="hardcover-fiction"
      >
        <div
          className="categorieImageBox"
        >
          <img
            alt="Categorie Icon"
            src=""
          />
        </div>
        <div>
          <h2>
            Hardcover Fiction
          </h2>
          <p>
            Published date:
            2023-03-19
          </p>
        </div>
      </section>
    </a>
  </section>,
]
`;
